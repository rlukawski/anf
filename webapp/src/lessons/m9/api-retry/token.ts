import axios from 'axios';
import axiosRetry from 'axios-retry';
import * as rax from 'retry-axios';
import { BASE_URL } from 'env/config';
import { isUnauthorizedError } from '../error-ui/messages/error-guards';
import { retryHTTP } from '../retryHTTP';

/**
 * UWAGA! ðŸ”¥
 *
 * W naszym repo szkoleniowym mamy osobny folder webapp/src/api, w ktÃ³rym sÄ… funkcje HTTP bez retry, a tu jest wersja z retry, bo dodaliÅ›my jÄ… w tym module.
 * Normalnie raczej wszystkie API trzymalibyÅ›my w 1 miejscu dla porzÄ…dku. Natomiast chcemy tutaj bardziej uwydatniÄ‡ elementy _dodane_ do szkolenia w konkretnym module.
 */

const client = axios.create({ baseURL: BASE_URL });

// Alternatywnie moÅ¼emy skorzystaÄ‡Â z gotowego pluginu do ponawiania zapytaÅ„ (https://www.npmjs.com/package/axios-retry)
// axiosRetry(client, { retries: 3 });
// axiosRetry(client, {
//   retries: 3,
//   retryDelay: axiosRetry.exponentialDelay
// });

// albo jeszcze innego (https://www.npmjs.com/package/retry-axios)
// client.defaults.raxConfig = {
//   instance: client,
//   retry: 3,
//   noResponseRetries: 3,
//   backoffType: 'exponential',
//   httpMethodsToRetry: ['GET', 'HEAD', 'OPTIONS', 'DELETE', 'PUT', 'POST'],
//   statusCodesToRetry: [[100, 199], [400, 499], [500, 599]],
// };
// rax.attach(client);

// Odkomentuj poniÅ¼sze, jeÅ¼eli uÅ¼ywasz axiosRetry albo rax
// export const getTokenInstruction = async () => {
//   const response = await client.post<TokenInstruction>(`/banking/token`)
//   return response.data
// }

export interface TokenInstruction {
  instruction: string
  tokenId: string
}

export const getTokenInstruction = retryHTTP(async () => {
  const response = await client.post<TokenInstruction>(`/banking/token`)
  return response.data
}, {
  maxAttempts: 3,
  backoff: () => 0
  // backoff: (attempt) => 1000 * attempt
})

export interface TokenConfirmation {
  tokenId: string
  tokenCode: string
}

export const sendTokenCode = retryHTTP(async (params: TokenConfirmation) => {
  const response = await client.post<void>(`/banking/token/${params.tokenId}`, params)
  return response.data
}, {
  maxAttempts: 3,
  backoff: () => 0,
  // backoff: (attempt) => 1000 * attempt,
  retryIf: (e) => {
    return !isUnauthorizedError(e)
  }
})
