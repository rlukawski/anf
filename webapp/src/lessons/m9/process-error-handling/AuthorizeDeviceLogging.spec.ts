import { AuthorizeDeviceWithProvidersPO } from "./AuthorizeDeviceWithProviders.po";

import { tokenMockHandlers } from 'api/mock/token.mock'
import { setupMockServer } from "api/mock/mock.server";
import { MockLogger } from "../logger/Logger.mock";

describe('AuthorizeDevice (Logging)', () => {
  setupMockServer(...tokenMockHandlers)

  it('should stop the process and logout after the user clicks logout', async () => {
    // given
    const mockLogger = new MockLogger()
    const authorizeDevicePO = AuthorizeDeviceWithProvidersPO.renderWithProviders(mockLogger)

    // then
    await authorizeDevicePO.expectTextDisplayed("Nieznane urządzenie")

    // when
    authorizeDevicePO.clickLogoutButton()

    // then
    mockLogger.expectLogs.toMatchSnapshot()
  });

  it('should cancel the authorization choice if the user clicks cancel on "add device"', async () => {
    // given
    const mockLogger = new MockLogger()
    const authorizeDevicePO = AuthorizeDeviceWithProvidersPO.renderWithProviders(mockLogger)

    // then
    await authorizeDevicePO.expectTextDisplayed("Nieznane urządzenie")

    // when
    authorizeDevicePO.clickChooseAddDeviceButton()

    // then
    await authorizeDevicePO.expectTextDisplayed("Zapisz to urządzenie jako zaufane")

    // when
    authorizeDevicePO.confirmAddDeviceName()

    // then
    await authorizeDevicePO.expectTextDisplayed("Zapisz to urządzenie jako zaufane")

    // when
    authorizeDevicePO.clickTokenCancelButton()

    // then
    await authorizeDevicePO.expectTextDisplayed("Nieznane urządzenie")

    // then
    mockLogger.expectLogs.toMatchSnapshot()
  });

  it('should authorize a device successfully after the user passes the correct password', async () => {
    // given
    const mockLogger = new MockLogger()
    const authorizeDevicePO = AuthorizeDeviceWithProvidersPO.renderWithProviders(mockLogger)

    // when
    authorizeDevicePO.clickChooseAddDeviceButton()

    // then
    await authorizeDevicePO.expectTextDisplayed("Zapisz to urządzenie jako zaufane")

    // when
    authorizeDevicePO.confirmAddDeviceName("mój komputerek")

    // then
    await authorizeDevicePO.expectTextDisplayed("mój komputerek")
    await authorizeDevicePO.expectTextDisplayed("Zapisz to urządzenie jako zaufane")

    // when
    authorizeDevicePO.submitAddDeviceToken("pass")

    // then
    await authorizeDevicePO.expectTextDisplayed("Urządzenie zapisane jako zaufane")

    // when
    authorizeDevicePO.clickCloseButton()

    // then
    mockLogger.expectLogs.toMatchSnapshot()
  });

  it('should block the process if the user passes an invalid password', async () => {
    // given
    const mockLogger = new MockLogger()
    const authorizeDevicePO = AuthorizeDeviceWithProvidersPO.renderWithProviders(mockLogger)

    // when
    authorizeDevicePO.clickChooseAddDeviceButton()

    // then
    await authorizeDevicePO.expectTextDisplayed("Zapisz to urządzenie jako zaufane")

    // when
    authorizeDevicePO.confirmAddDeviceName()

    // then
    await authorizeDevicePO.expectTextDisplayed("Zapisz to urządzenie jako zaufane")

    // when
    authorizeDevicePO.submitAddDeviceToken("pass too long")

    // then
    await authorizeDevicePO.expectTextDisplayed("Niepoprawny token")

    // then
    mockLogger.expectLogs.toMatchSnapshot()
  });

  it('should allow once after the user passes the correct password', async () => {
    // given
    const mockLogger = new MockLogger()
    const authorizeDevicePO = AuthorizeDeviceWithProvidersPO.renderWithProviders(mockLogger)

    // when
    authorizeDevicePO.clickChooseAllowOnceButton()

    // then
    await authorizeDevicePO.expectTextDisplayed("Jednorazowy wjazd do apki")

    // when
    authorizeDevicePO.submitAllowOnceToken("pass")

    // then
    await authorizeDevicePO.expectTextDisappeared("Jednorazowy wjazd do apki")
    await authorizeDevicePO.expectLoaderDisappeared()

    // then
    mockLogger.expectLogs.toMatchSnapshot()
  });
});
