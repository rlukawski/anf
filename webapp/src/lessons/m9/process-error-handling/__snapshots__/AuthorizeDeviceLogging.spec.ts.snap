// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorizeDevice (Logging) should allow once after the user passes the correct password 1`] = `
Array [
  Array [
    "DEBUG",
    "User has started process: authorize device",
  ],
  Array [
    "INFO",
    "User has clicked to allow once",
    undefined,
  ],
  Array [
    "INFO",
    "The device has been successfully authorized once",
    undefined,
  ],
]
`;

exports[`AuthorizeDevice (Logging) should authorize a device successfully after the user passes the correct password 1`] = `
Array [
  Array [
    "DEBUG",
    "User has started process: authorize device",
  ],
  Array [
    "INFO",
    "User has clicked to add a trusted device",
    undefined,
  ],
  Array [
    "INFO",
    "User has chosen the device name: mój komputerek",
    undefined,
  ],
  Array [
    "INFO",
    "The device has been successfully added to trusted devices",
    undefined,
  ],
  Array [
    "INFO",
    "User has successfully finished the process",
    undefined,
  ],
]
`;

exports[`AuthorizeDevice (Logging) should block the process if the user passes an invalid password 1`] = `
Array [
  Array [
    "DEBUG",
    "User has started process: authorize device",
  ],
  Array [
    "INFO",
    "User has clicked to add a trusted device",
    undefined,
  ],
  Array [
    "INFO",
    "User has chosen the device name: Mój komputer",
    undefined,
  ],
  Array [
    "INFO",
    "User has sent an invalid confirmation token",
    undefined,
  ],
]
`;

exports[`AuthorizeDevice (Logging) should cancel the authorization choice if the user clicks cancel on "add device" 1`] = `
Array [
  Array [
    "DEBUG",
    "User has started process: authorize device",
  ],
  Array [
    "INFO",
    "User has clicked to add a trusted device",
    undefined,
  ],
  Array [
    "INFO",
    "User has chosen the device name: Mój komputer",
    undefined,
  ],
  Array [
    "INFO",
    "User has cancelled the authorization method choice",
    undefined,
  ],
]
`;

exports[`AuthorizeDevice (Logging) should stop the process and logout after the user clicks logout 1`] = `
Array [
  Array [
    "DEBUG",
    "User has started process: authorize device",
  ],
  Array [
    "INFO",
    "User has cancelled authorizing the device",
    undefined,
  ],
]
`;
